pkg/sql/alter_column_type.go

```diff
@@ -0,0 +1,239 @@
+					"you can enable alter column type general support by running `SET enable_experimental_alter_column_type_general = true`")
+		}
+		if err := alterColumnTypeGeneral(tableDesc, col, t, params); err != nil {
+			return err
+		}
+		if err := params.p.createOrUpdateSchemaChangeJob(params.ctx, tableDesc, "alter column type", tableDesc.ClusterVersion.NextMutationID); err != nil {
+			return err
+		}
+		params.p.SendClientNotice(params.ctx, errors.New("alter column type changes are finalized asynchronously; "+
+			"further schema changes on this table may be restricted until the job completes; "+
+			"some inserts into the altered column may not be supported until the schema change is finalized"))
+	default:
+		return fmt.Errorf("unknown conversion for %s -> %s",
+			col.Type.SQLString(), typ.SQLString())
+	}
+
+	return nil
+}
+
+func alterColumnTypeGeneral(
+	tableDesc *sqlbase.MutableTableDescriptor,
+	col *sqlbase.ColumnDescriptor,
+	t *tree.AlterTableAlterColumnType,
+	params runParams,
+) error {
+	// Disallow ALTER COLUMN ... TYPE ... USING EXPRESSION.
+	// Todo(richardjcai): Todo, need to handle "inverse" expression
+	// during state after column swap but the old column has not been dropped.
+	// Can allow the user to provide an inverse expression.
+	if t.Using != nil {
+		return UsingExpressionNotSupportedErr
+	}
+
+	// Disallow ALTER COLUMN TYPE general for columns that are
+	// part of indexes.
+	for i := range tableDesc.Indexes {
+		for _, id := range append(
+			tableDesc.Indexes[i].ColumnIDs,
+			tableDesc.Indexes[i].ExtraColumnIDs...) {
+			if col.ID == id {
+				return ColInIndexNotSupportedErr
+			}
+		}
+	}
+
+	for _, id := range append(
+		tableDesc.PrimaryIndex.ColumnIDs,
+		tableDesc.PrimaryIndex.ExtraColumnIDs...) {
+		if col.ID == id {
+			return ColInIndexNotSupportedErr
+		}
+	}
+
+	currentMutationID := tableDesc.ClusterVersion.NextMutationID
+	for i := range tableDesc.Mutations {
+		mut := &tableDesc.Mutations[i]
+		if mut.MutationID < currentMutationID {
+			return unimplemented.NewWithIssuef(
+				47137, "table %s is currently undergoing a schema change", tableDesc.Name)
+		}
+	}
+
+	nameExists := func(name string) bool {
+		_, _, err := tableDesc.FindColumnByName(tree.Name(name))
+		return err == nil
+	}
+
+	shadowColName := sqlbase.GenerateUniqueConstraintName(col.Name, nameExists)
+
+	// The default computed expression is casting the column to the new type.
+	newComputedExpr := tree.CastExpr{
+		Expr:       &tree.ColumnItem{ColumnName: tree.Name(col.Name)},
+		Type:       t.ToType,
+		SyntaxMode: tree.CastShort,
+	}
+	s := tree.Serialize(&newComputedExpr)
+	newColComputeExpr := &s
+
+	// Create the default expression for the new column.
+	hasDefault := col.HasDefault()
+	var newColDefaultExpr *string
+	if hasDefault {
+		if col.HasNullDefault() {
+			s := tree.Serialize(tree.DNull)
+			newColDefaultExpr = &s
+		} else {
+			// The "default" default expression for the new column is casting the
```

Comment:
what does the second default mean here?

Reply:
I meant the default expression for the column by default is casting the old default column to the new type. May be different if the user provides an expression for the schema change